PR for optimizing T count for all states in qasm format

typerSniper opened the pull request:
The function `opt:: ([ID] -> [ID] -> [Primitive] -> [Primitive])` takes a set of qubits, a set of inputs, and a circuit representation, and returns an optimized circuit. When called for qasm inputs, I ensure that the set of inputs is equal to the set of qubits, thus treating all of the qubits equally in an unknown state and directing the optimization to work for all inputs.


If you think this looks good, I will go ahead and add a flag named `-nz` ("no zeroes") that decides between the two alternatives.

meamy commented:
That should do the trick. Please do add the flag and I'll test it out. To maintain consistency with the other flags please use a descriptive name, e.g. -purecircuit to denote that the qasm code is to be viewed as defining a circuit rather than a program. -uninitializedregisters I suppose is also an option.

typerSniper commented:
Ok, very nice. I have added the flag `purecircuit` and also added a description for it in the `printHelp` section.  The changes only kick in when the flag is given. Let me know what you think!

-Jatin

meamy commented:
Merged, thanks!
